jax/experimental/mosaic/gpu/examples/fdgg.py

```diff
@@ -0,0 +1,505 @@
+    return arith.divui(arith.subi(arith.addi(x, y),
+                                  cx(1)),
+                       y)
+
+  allocator = WarpAllocator()
+  warp = allocator.alloc_warp
+  warpgroup = allocator.alloc_warpgroup
+
+  tma_warp = warp()
+  mma_warp = warp()
+
+
+  def fdgg(ctx, a, w, group_sizes, d, smem):
+    ((a_smem, b_smem), d_smem), barriers, mma_done_barrier, acc = smem
+    (ab_full_barriers, ab_empty_barriers) = barriers
+
+    warp_idx = arith.index_cast(index, mgpu.warp_idx(sync=True))
+
+    @contextlib.contextmanager
+    def only(w):
+      with _only(w, warp_idx):
+        yield
+
+    worker_id = gpu.block_id(gpu.Dimension.x)
+    worker_count = cx(cta_count)
+    work_id_group_start = cx(0)
+    initial_work_id = worker_id
+    group_offset = cx(0)
+    @mgpu.fori(cx(expert_count), [initial_work_id, work_id_group_start, group_offset])
+    def group_for_body(expert_id, carrys):
+      work_id, work_id_group_start, group_offset = carrys
+      group_size = arith.index_cast(index,
+                                    memref.load(group_sizes, [expert_id]))
+      group_work_item_count = arith.muli(ceildiv(group_size, cx(tile_m)),
+                                         cx(n_tile_count))
+      work_id_group_end = arith.addi(work_id_group_start, group_work_item_count)
+
+      @subgroup_for(work_id, work_id_group_end, worker_count)
+      def subgroup_for_body(work_id):
+        tile_n_start = arith.muli(arith.remui(work_id, cx(n_tile_count)),
+                                  cx(tile_n))
+        work_id_within_group = arith.subi(work_id, work_id_group_start)
+        subgroup = arith.divui(work_id_within_group, cx(n_tile_count))
+        # TODO(andportnoy) try to pick better names for all these variables
+        #                  and streamline layout calculations in general
+        tile_m_start_within_group = arith.muli(subgroup, cx(tile_m))
+        tile_m_start = arith.addi(group_offset, tile_m_start_within_group)
+        tile_m_end_within_group = arith.minui(arith.addi(tile_m_start_within_group, cx(tile_m)),
+                                              group_size)
+        tile_m_end = arith.addi(group_offset, tile_m_end_within_group)
+        tile_m_count = arith.subi(tile_m_end, tile_m_start)
+
+        # output n_start
+        o_n_start = tile_n_start
+
+        # weight n_start
+        w_n_start = arith.addi(arith.muli(expert_id, cx(expert_n)),
+                               tile_n_start)
+
+        local_work_id = arith.divui(work_id, worker_count)
+        tmem_slot = arith.remui(local_work_id, cx(tmem_slot_count))
+
+        tmem_address = arith.addi(acc.address,
+                                  arith.index_cast(i32,
+                                                   arith.muli(tmem_slot, cx(tile_n))))
+        acc_slot = tcgen05.TMEMRef(
+            address=tmem_address,
+            shape=(tile_m, tile_n),
+            layout=acc.layout,
+            dtype=acc.dtype,
+        )
```

Comment:
This is just slicing the TMEM ref, right? Perhaps extend `TMEMRef.slice` to support this, instead?